python/nx-cugraph/nx_cugraph/algorithms/centrality/katz.py

```diff
@@ -0,0 +1,98 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import networkx as nx
+import numpy as np
+import pylibcugraph as plc
+
+from nx_cugraph.convert import _to_graph
+from nx_cugraph.utils import (
+    _dtype_param,
+    _get_float_dtype,
+    networkx_algorithm,
+    not_implemented_for,
+)
+
+__all__ = ["katz_centrality"]
+
+
+@not_implemented_for("multigraph")
+@networkx_algorithm(extra_params=_dtype_param)
+def katz_centrality(
+    G,
+    alpha=0.1,
+    beta=1.0,
+    max_iter=1000,
+    tol=1.0e-6,
+    nstart=None,
+    normalized=True,
+    weight=None,
+    *,
+    dtype=None,
+):
+    """`nstart` parameter is not used, and `normalized=False` is not supported."""
+    if not normalized:
+        raise NotImplementedError("normalized=False is not supported.")
+    G = _to_graph(G, weight, np.float32)
+    if (N := len(G)) == 0:
+        return {}
+    if dtype is not None:
+        dtype = _get_float_dtype(dtype)
+    elif weight in G.edge_values:
+        dtype = _get_float_dtype(G.edge_values[weight].dtype)
+    else:
+        dtype = np.float32
+    if nstart is not None:
+        # Check if given nstart is valid even though we don't use it
+        nstart = G._dict_to_nodearray(nstart, 0, dtype)
+    b = bs = None
+    try:
+        b = float(beta)
+    except (TypeError, ValueError) as exc:
+        try:
+            bs = G._dict_to_nodearray(beta, dtype=dtype)
+            b = 1.0  # float value must be given to PLC (and will be ignored)
+        except (KeyError, ValueError):
+            raise nx.NetworkXError(
+                "beta dictionary must have a value for every node"
+            ) from exc
+    try:
+        node_ids, values = plc.katz_centrality(
+            resource_handle=plc.ResourceHandle(),
+            graph=G._get_plc_graph(weight, 1, dtype, store_transposed=True),
+            betas=bs,
+            alpha=alpha,
+            beta=b,
+            epsilon=N * tol,
+            max_iterations=max_iter,
+            do_expensive_check=False,
+        )
+    except RuntimeError as exc:
+        # Errors from PLC are sometimes a little scary and not very helpful
+        raise nx.PowerIterationFailedConvergence(max_iter) from exc
+    return G._nodearrays_to_dict(node_ids, values)
+
+
+@katz_centrality._can_run
+def _(
+    G,
+    alpha=0.1,
+    beta=1.0,
+    max_iter=1000,
+    tol=1.0e-6,
+    nstart=None,
+    normalized=True,
+    weight=None,
+    *,
+    dtype=None,
+):
+    return normalized
```

Comment:
Can you remind me again how `_can_run` is used? I'm mainly wondering if the `NotImplementedError` being raised above is redundant or if it's needed for something `_can_run` doesn't handle?

Also, is there a reason this isn't checking `nstart` and returning False if it's set?  Same question for eigenvector_centrality.  I think I'm not understanding the difference between "unused" and "unsupported" from the user's perspective.

Reply:
It's run when going through the networkx dispatch _before_ input data is converted. We shouldn't make assumptions about the input arguments, and we shouldn't convert the graphs in this function.

Raising `NotImplementedError` when `can_run` will return False is indeed redundant. It is included in case the function is called directly and not via dispatching.

> Also, is there a reason this isn't checking nstart and returning False if it's set? Same question for eigenvector_centrality. I think I'm not understanding the difference between "unused" and "unsupported" from the user's perspective.

I consider `nstart` to be an optimization. Generally speaking, it shouldn't change the answer, but it may change the number of iterations required. So, these algorithms run just fine w/o `nstart`, hence why `can_run` still returns True if nstart is given.

Reply:
Contrast this with `dangling` in pagerank, which _does_ generally change the result.

Reply:
I see what you mean re: optimizations vs. args that change results, so for this check ("can this run at all or not?") this makes sense.